libcfnet/file_stream.c

```diff
@@ -0,0 +1,943 @@
+
+    /* Set message length */
+    uint16_t header = len << 4;
+
+    /* Set Error flag */
+    if (err)
+    {
+        header |= (1 << 0);
+    }
+
+    /* Set End-of-File flag */
+    if (eof)
+    {
+        header |= (1 << 1);
+    }
+
+    /* Send header */
+    header = htons(header);
+    int ret = TLSSend(conn, (char *) &header, HEADER_SIZE);
+    if (ret != HEADER_SIZE)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to send message header during file stream (%d != %d)",
+            ret,
+            HEADER_SIZE);
+        return false;
+    }
+
+    if (len > 0)
+    {
+        /* Send payload */
+        ret = TLSSend(conn, msg, len);
+        if (ret != (int) len)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Failed to send message payload during file stream (%d != %zu)",
+                ret,
+                len);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * @brief Send a message using the file stream protocol
+ *
+ * @param conn The SSL connection object
+ * @param msg The message to send
+ * @param len The length of the message to send (must be less or equal to
+ *            MESSAGE_SIZE Bytes)
+ * @param eof Set to true if this is the last message in a transaction,
+ *            otherwise false
+ * @return true on success, otherwise false
+ */
+static inline bool SendMessage(
+    SSL *conn, const char *msg, size_t len, bool eof)
+{
+    assert(conn != NULL);
+    assert(msg != NULL || len == 0);
+
+    return __SendMessage(conn, msg, len, eof, false);
+}
+
+/**
+ * @brief Receive a message using the file stream protocol
+ *
+ * @param conn The SSL connection object
+ * @param msg The message receive buffer (must be MESSAGE_SIZE Bytes large)
+ * @param len The length of the reveived message
+ * @param eof Is set to true if this was the last message in the transaction
+ * @return true on success, otherwise false
+ *
+ * @note RecvMessage fails if the communication is broken or if we received an
+ *       error from the remote host. In both cases, we should not try to flush
+ *       the stream.
+ */
+static bool RecvMessage(SSL *conn, char *msg, size_t *len, bool *eof)
+{
+    assert(conn != NULL);
+    assert(msg != NULL);
+    assert(len != NULL);
+    assert(eof != NULL);
+
+    /* TLSRecv() expects a buffer this size */
+    char recv_buffer[CF_BUFSIZE];
+
+    /* Receive header */
+    uint16_t header;
+    int ret = TLSRecv(conn, recv_buffer, HEADER_SIZE);
+    if (ret != HEADER_SIZE)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to receive message header during file stream (%d != %d)",
+            ret,
+            HEADER_SIZE);
+        return false;
+    }
+    memcpy(&header, recv_buffer, HEADER_SIZE);
+    header = ntohs(header);
+
+    /* Extract Error flag */
+    bool err = header & (1 << 0);
+
+    /* Extract End-of-File flag */
+    *eof = header & (1 << 1);
+
+    /* Extract message length */
+    *len = header >> 4;
```

Comment:
This requires `len` (`size_t`) to be at least 3 bytes, right? Probably that's guaranteed, but I'd add a static assert for this to avoid potential nasty surprises.

Reply:
If it was left shift, then yes, 3 bytes. However, with right shift, two bytes should do. The following:
```C
    assert(sizeof(*len) >= 2); /* "It's probably guaranteed, but let's make
                                * sure to avoid potentially nasty surprises."
                                *  - @vpodzime */
```
still would not hurt.

Reply:
Right, it's 12 bits, not 16. I got confused by `uint16_t`.